index.js

```diff
@@ -1,29 +1,255 @@
+    this.PACKAGE = new Cache();
+    this.PATH = new Cache();
+    this.REAL_FILE_PATH = new Cache();
+  }
+}
+
+var TEMP_CACHES = {
+  PACKAGE: new Cache(),
+  PATH: new Cache(),
+  REAL_FILE_PATH: new Cache(),
+
+  reset: function() {
+    this.PACKAGE = new Cache();
+    this.PATH = new Cache();
+    this.REAL_FILE_PATH = new Cache();
+  }
+}
+
+/*
+ * Pointer to the set of caches currently in use. When skipCache is not true it is
+ * set to PERMANENT_CACHES. When it is true, this is set to TEMP_CACHES and the
+ * 'reset' function is called at the end of the hashForDep call.
+ */
+var CACHES;
+
+function setCaches(skipCache) {
+  if (skipCache) {
+    CACHES = TEMP_CACHES;
+    CACHES.reset();
+  } else {
+    CACHES = PERMANENT_CACHES;
+  }
+}
+
+/*
+ * Entries in 'PACKAGE' cache. The idea is to minimize the information needed for
+ * each entry (don't store the entire package.json object), but keep
+ * information around for all parts of computing a hash for the dependency
+ * tree of the original package named by 'name' and 'dir' in hashForDep.
+ * The PackageCacheEntry is created at the time the package.json file is
+ * read in (so we only have to do that once), but some properties
+ * may not be initialized until later (this then supports creating cycles
+ * in the dependencies list, which is supported by Node).
+ */
+var PackageCacheEntry = function(name, version, rootDir, sourceHash) {
+
+  // the resolved real file path for the package.json file,
+  // the same value as 'realPath' in genRealPathKey.
+  this.rootDir = rootDir;  
+
+  // the name of the package, from its package.json
+  this.name = name;
+
+  // the version of the package, from its package.json
+  this.version = version; 
+
+  // the computed hash for only the package's source files (under
+  // the root dir, but specifically NOT any node_modules dependencies).
+  this.sourceHash = sourceHash;
+
+  // the computed hash for the package's source files plus
+  // the source hashes of the dependencies. Because we compute
+  // dependencies as 'all the way down', this gets everything.
+  this.hash;
+
+  // references to other PackageCacheEntry objects that correspond
+  // to the dependencies listed in the package.json
+  this.dependencies = [];   
+}
 
 function HashForDepSchema() {
   this.paths = 0;
 }
 
-function cacheGet(key) {
-  return CACHE[key]
+function genRealPathKey(realPath) {
+  return cacheKey('hashed:' + realPath, '');
 }
 
-function cacheGet(key) {
-  return CACHE[key]
+function genNameDirKey(name, dir) {
+  return cacheKey(name, dir);
+}
+
+/*
+ * Resolve the real path for a file, return null if does not
+ * exist or is not a file or FIFO, return the real path otherwise.
+ *
+ * @private
+ * @method getRealFilePath
+ * @param  {String} filePath the path to the file of interest
+ * @return {String} real path or null
+ */
+function getRealFilePath(filePath) {
+  var realPath;
+
+  try {
+    realPath = CACHES.REAL_FILE_PATH.get(filePath);
```

Comment:
the cache check, likely should not be in the `try/catch`, I believe the `try/Catch` is focused on the IO/FS related errors.

Reply:
Moved setting realPath and the check for non-null out of the try/catch